gnucash/gnome-utils/gnc-amount-edit.c

```diff
@@ -120,22 +123,87 @@ gnc_amount_edit_init (GNCAmountEdit *gae)
     gae->evaluate_on_enter = FALSE;
+    gae->block_changed = FALSE;
 
     // Set the name for this widget so it can be easily manipulated with css
     gtk_widget_set_name (GTK_WIDGET(gae), "gnc-id-amount-edit");
 
     g_signal_connect (G_OBJECT(gae), "changed",
-                      G_CALLBACK(gnc_amount_edit_changed), NULL);
+                      G_CALLBACK(gnc_amount_edit_changed), gae);
+
+    g_signal_connect (G_OBJECT(gae), "paste-clipboard",
+                      G_CALLBACK(gnc_amount_edit_paste_clipboard), NULL);
 }
 
 static void
 gnc_amount_edit_changed (GtkEditable *editable, gpointer user_data)
 {
-    /*GTK_EDITABLE_CLASS(parent_class)->changed(editable);*/
+    GNCAmountEdit *gae = GNC_AMOUNT_EDIT(user_data);
+     /*GTK_EDITABLE_CLASS(parent_class)->changed(editable);*/
+    if (gae->block_changed)
+        g_signal_stop_emission_by_name (G_OBJECT(gae), "changed");
 
     GNC_AMOUNT_EDIT(editable)->need_to_parse = TRUE;
 }
 
+static void
+gnc_amount_edit_paste_clipboard (GNCAmountEdit *gae, gpointer user_data)
+{
+    GtkClipboard *clipboard = gtk_widget_get_clipboard (GTK_WIDGET(gae),
+                                                        GDK_SELECTION_CLIPBOARD);
+    gchar *text = gtk_clipboard_wait_for_text (clipboard);
+
+    if (text)
+    {
+        gchar *filtered_text = gnc_filter_text_for_control_chars (text);
+        gchar *existing = gtk_editable_get_chars (GTK_EDITABLE(gae), 0, -1);
+        gint start_pos, end_pos;
+        gint position;
+
+        if (!filtered_text)
+        {
+            g_free (text);
+            g_free (existing);
+            return;
+        }
+
+        position = gtk_editable_get_position (GTK_EDITABLE(gae));
+
+        if (gtk_editable_get_selection_bounds (GTK_EDITABLE(gae),
+                                               &start_pos, &end_pos))
+        {
+            gint old_len = g_utf8_strlen (existing, -1);
+            gchar *begin = g_utf8_substring (existing, 0, start_pos);
+            gchar *end = g_utf8_substring (existing, end_pos, old_len);
+            gchar *changed_text = g_strdup_printf ("%s%s%s", begin, filtered_text, end);
+
+            gae->block_changed = TRUE;
+            gtk_editable_delete_text (GTK_EDITABLE(gae), 0, -1);
+            gae->block_changed = FALSE;
+
+            gtk_editable_insert_text (GTK_EDITABLE(gae),
+                                      changed_text, -1, &position);
+
+            position = start_pos + g_utf8_strlen (filtered_text, -1);
+
+            g_free (begin);
+            g_free (end);
+            g_free (changed_text);
+        }
+        else
+            gtk_editable_insert_text (GTK_EDITABLE(gae),
+                                      filtered_text, -1, &position);
+
+        gtk_editable_set_position (GTK_EDITABLE(gae), position);
+
+        g_signal_stop_emission_by_name (G_OBJECT(gae), "paste-clipboard");
+
+        g_free (text);
+        g_free (existing);
+        g_free (filtered_text);
+    }
+}
```

Comment:
Argh, another copy of this code. Since this isn't in register the single function should be in app-utils/gnc-ui-utils.c with `gnc_filter_text*`.